### ISSUE TITLE
reimplement fcn-xs example using gluon

### ISSUE THREAD

**Issue by qiliux** [apache/incubator-mxnet#8396]
Package used (Python/R/Scala/Julia):
(I'm using ...)

Python
My problem is that how to use the pre-layer's info in  my custom layer to implement skip architecture. 
```
class MyLayer(Block):
    def __init__(self, **kwargs):
        super(MyLayer, self).__init__(**kwargs)

    def forward(self, x):
        I want to use the second Pool layer info here.
        to Reimplemnt FCN-xs in the example use gluon.

net = gluon.nn.Sequential()
with net.name_scope():
    net.add(gluon.nn.Conv2D(channels=20, kernel_size=3, activation='relu'))
    net.add(gluon.nn.MaxPool2D(pool_size=2, strides=2))
    net.add(gluon.nn.Conv2D(channels=50, kernel_size=3, activation='relu'))
    net.add(gluon.nn.MaxPool2D(pool_size=2, strides=2))
    # The Flatten layer collapses all axis, except the first one, into one axis.
    net.add(gluon.nn.Flatten())
    net.add(gluon.nn.Dense(num_fc, activation="relu"))
    net.add(MyLayer(Block)) # use the MaxPool2D info.
    net.add(gluon.nn.Dense(num_outputs))
```
## What have you tried to solve it?

1.  See the gluon network souce code
2. google it

**Comment by piiswrong** [comment]
You can implement the whole thing in a custom block, without using sequential

**Comment by qiliux** [comment]
@piiswrong thank you. I will try it. If i succeed, i will update the answer.

**Comment by qiliux** [comment]
@piiswrong can you help me check if my thought is right? I try my own dataset, but i can't make the fcn work. The loss is always high.
And i met a problem: The net output is N\*C\*H\*W. But my data label is N\*1\*H\*W. `loss = mx.ndarray.SoftmaxOutput(output, label, multi_output=True)` won't work, if i don't add `multi_output=True`

```
from mxnet.gluon import nn

class FCN_VGG16(gluon.HybridBlock):
    def __init__(self, num_outputs, **kwargs):
        
        super(FCN_VGG16, self).__init__(**kwargs)
        self.num_outputs=num_outputs
        with self.name_scope():
            # VGG16 
            self.conv1_1 = nn.Conv2D(channels=64, kernel_size=(3,3), padding=100,activation="relu")
            self.conv1_2 = nn.Conv2D(channels=64, kernel_size=(3,3), padding=1,activation="relu")
            self.pool1 = nn.MaxPool2D(pool_size=2,strides=2)
            
            self.conv2_1 = nn.Conv2D(channels=128, kernel_size=(3,3), padding=1,activation="relu")
            self.conv2_2 = nn.Conv2D(channels=128, kernel_size=(3,3), padding=1,activation="relu")
            self.pool2 = nn.MaxPool2D(pool_size=2,strides=2)
            
            self.conv3_1 = nn.Conv2D(channels=256, kernel_size=(3,3), padding=1,activation="relu")
            self.conv3_2 = nn.Conv2D(channels=256, kernel_size=(3,3), padding=1,activation="relu")
            self.conv3_3 = nn.Conv2D(channels=256, kernel_size=(3,3), padding=1,activation="relu")
            
            self.pool3 = nn.MaxPool2D(pool_size=2,strides=2)
            self.conv4_1 = nn.Conv2D(channels=512, kernel_size=(3,3), padding=1,activation="relu")
            self.conv4_2 = nn.Conv2D(channels=512, kernel_size=(3,3), padding=1,activation="relu")
            self.conv4_3 = nn.Conv2D(channels=512, kernel_size=(3,3), padding=1,activation="relu")
            self.pool4 = nn.MaxPool2D(pool_size=2,strides=2)
            
            self.conv5_1 = nn.Conv2D(channels=512, kernel_size=(3,3), padding=1,activation="relu")
            self.conv5_2 = nn.Conv2D(channels=512, kernel_size=(3,3), padding=1,activation="relu")
            self.conv5_3 = nn.Conv2D(channels=512, kernel_size=(3,3), padding=1,activation="relu")
            self.pool5 = nn.MaxPool2D(pool_size=2,strides=2)

            self.conv6 = nn.Conv2D(channels=4096, kernel_size=(7,7),activation="relu")
            self.drop1 = nn.Dropout(.5)
            
            self.conv7 = nn.Conv2D(channels=4096, kernel_size=(1,1),activation="relu")
            self.drop2 = nn.Dropout(.5)
            
            self.conv8 = nn.Conv2D(channels=num_outputs, kernel_size=(1,1),activation="relu")
            
            self.deconv = nn.Conv2DTranspose(channels=num_outputs,kernel_size=(64,64),strides=(32,32))
            
    def hybrid_forward(self, F, x):
        #print('type(x): {}, F: {}'.format(
         #   type(x).__name__, F.__name__))
        #print(x.shape)

        x = self.conv1_1(x)
        x = self.conv1_2(x)
        x = self.pool1(x)
        x = self.conv2_1(x)
        x = self.conv2_2(x)
        x = self.pool2(x)
        x = self.conv3_1(x)
        x = self.conv3_2(x)
        x = self.conv3_3(x)
        x = self.pool3(x)
        x = self.conv4_1(x)
        x = self.conv4_2(x)
        x = self.conv4_3(x)
        x = self.pool4(x)
        x = self.conv5_1(x)
        x = self.conv5_2(x)
        x = self.conv5_3(x)
        x = self.pool5(x)

        x = self.conv6(x)
        x = self.drop1(x)
        x = self.conv7(x)
        x = self.drop2(x)
        x = self.conv8(x)
        x = self.deconv(x)
        # How to use softmax and Crop here?
#         x = F.slice_axis(x,axis=2,begin=12,end=372)
#         x = F.slice_axis(x,axis=3,begin=12,end=372)
#         x = F.softmax(data=x)
        #print("FF-shape: ",x.shape)
        return x
net = FCN_VGG16(2)
# net.hybridize()
#print(net.collect_params())
net.collect_params().initialize(mx.init.Xavier(magnitude=1.24),ctx=ctx)

trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': 1e-7,'momentum':0.99, 'wd': 0.005})
loss = gluon.loss.SoftmaxCrossEntropyLoss()
epochs = 10
smoothing_constant = .01

for e in range(epochs):
    for i, (data,label) in enumerate(train_data):
        with autograd.record():
            data = data.as_in_context(ctx)
            label = label.as_in_context(ctx)

            output = net(data)
            # Do i use loss function right?
            loss = mx.ndarray.SoftmaxOutput(output, label, multi_output=True)

        loss.backward()
        trainer.step(data.shape[0])
        curr_loss = nd.mean(loss).asscalar()
        moving_loss = (curr_loss if ((i == 0) and (e == 0))
                       else (1 - smoothing_constant) * moving_loss + smoothing_constant * curr_loss)
   
    print("Epoch %s. Loss: %s" % (e, moving_loss))
```

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.

**Comment by leleamol** [comment]
Recommended Labels: "Question","Gluon"

**Comment by ThomasDelteil** [comment]
@qiliux there is gluon implementation of the FCN available on [GluonCV](https://gluon-cv.mxnet.io/)
- See the tutorial [here](https://gluon-cv.mxnet.io/build/examples_segmentation/train_fcn.html#sphx-glr-build-examples-segmentation-train-fcn-py)
- See the code [here](https://github.com/dmlc/gluon-cv/blob/master/gluoncv/model_zoo/fcn.py)

@yzhliu can you please close?
@qiliux if you would like to follow up, please create a post on https://discuss.mxnet.io, thanks!
